src/H5Centry.c

```diff
@@ -594,8 +594,11 @@ H5C__flush_single_entry(H5F_t *f, H5C_cache_entry_t *entry_ptr, unsigned flags)
                 else
                     mem_type = entry_ptr->type->mem_type;
 
-                if (H5F_block_write(f, mem_type, entry_ptr->addr, entry_ptr->size, entry_ptr->image_ptr) < 0)
+                if (H5F_block_write(f, mem_type, entry_ptr->addr, entry_ptr->size, entry_ptr->image_ptr) < 0) {
+                    H5C__REMOVE_ENTRY_FROM_SLIST(cache_ptr, entry_ptr, during_flush, FAIL);
```

Comment:
This seems like something that should be handled by the caller rather than the function itself.

Reply:
There are likely other things to be considered here, such as:

- the entry may also need to be removed from the collective list when operating in parallel
- the entry may also need to go through the rest of the `destroy` sequence even on write failure, unless it should also be left up to the caller to handle that

Reply:
Agree

Reply:
Also - how is this change to writing cache entries related to the crash on reading?

Reply:
The crash happened during the shutdown process.  When some memory was not freed, causing failure because the factory list still have allocated objects, and processing the failure crashed:
H5Eint.c:1464: H5E__push_stack: Assertion `cls_id > 0' failed.

This shows how the writing cache entries got involved:
  ...
  #013: ../../src/H5Fint.c line 1538 in H5F__dest(): problems closing file
    major: File accessibility
    minor: Unable to release object
  #014: ../../src/H5AC.c line 477 in H5AC_dest(): can't destroy cache
    major: Object cache
    minor: Unable to free object
  #015: ../../src/H5C.c line 546 in H5C_dest(): disable slist on flush dest failure failed
    major: Object cache
    minor: Internal error detected
  #016: ../../src/H5C.c line 1115 in H5C_set_slist_enabled(): slist not empty?
    major: Object cache
    minor: Internal error detected
  #017: ../../src/H5C.c line 502 in H5C_dest(): unable to flush cache
    major: Object cache
    minor: Unable to flush data from cache
  #018: ../../src/H5Cint.c line 1052 in H5C__flush_invalidate_cache(): flush invalidate ring failed
    major: Object cache
    minor: Unable to flush data from cache
  #019: ../../src/H5Cint.c line 1349 in H5C__flush_invalidate_ring(): dirty entry flush destroy failed
    major: Object cache
    minor: Unable to flush data from cache
  #020: ../../src/H5Centry.c line 598 in H5C__flush_single_entry(): Can't write image to file
    major: Object cache
    minor: Unable to flush data from cache
  #021: ../../src/H5Fio.c line 219 in H5F_block_write(): write through page buffer failed
    major: Low-level I/O
    minor: Write failed
  #022: ../../src/H5PB.c line 1001 in H5PB_write(): write through metadata accumulator failed
    major: Page Buffering
    minor: Write failed
  #023: ../../src/H5Faccum.c line 633 in H5F__accum_write(): file write failed
    major: Low-level I/O
    minor: Write failed
  #024: ../../src/H5FDint.c line 312 in H5FD_write(): addr overflow, addr = 4294967295, size=16, eoa=2848
    major: Invalid arguments to routine
    minor: Address overflowed
H5Fclose failed

I'd appreciate any suggestions.

Reply:
> There are likely other things to be considered here, such as:
> 
> * the entry may also need to be removed from the collective list when operating in parallel
> * the entry may also need to go through the rest of the `destroy` sequence even on write failure, unless it should also be left up to the caller to handle that

I'm looking into these suggestions

Reply:
I guess my suggestion is to find out what is happening when the read occurs that causes memory to be corrupted (which is almost certainly why the eventual crash occurs).

Reply:
The corrupted data was acquired in H5HL__hdr_deserialize() at:

H5_DECODE_LENGTH_LEN(image, heap->dblk_size, udata->sizeof_size);                             H5_DECODE_LENGTH_LEN(image, heap->free_block, udata->sizeof_size);

during the call to H5Gcreate1().  Both heap->dblk_size and heap->free_block are huge.  I put in the fix (NF's help) to catch that but during the shutdown process, memory leaks caused subsequent crashes, hence, the need to close those.

Although, at this point, I just realized I missed something in the PR, let me take care of that while you're thinking. :-)